### DISCUSSION TITLE
How to install?

### DISCUSSION THREAD

**Discussion by freemin7** [plasma-umass/coz#202] [Q&A] [answered]
I've tried to run some of the `coz` examples.  Here is how i what i tried so far (this is both for helping me think through what i am doing and help maintainers get an idea of the first time user experience) :
### Use Ubuntu's package and try to build the examples. 
`sudo apt install coz-profiler` works. I receive a `/usr/bin/coz` without any version info with the md5sum `0b6202d11cdfa6a35b9b7bd24917434c` which differs from the latest version https://raw.githubusercontent.com/plasma-umass/coz/381cbe57f2b47e8a8ca8a010978a09343246a4ec/coz . 
```
find / -name 'pca' 2> /dev/null
find / -name 'producer*' 2> /dev/null
```
Have no relevant results so the Ubuntu version does not ship the examples. I find no `coz-profiler-extras` in the registry. This is fine.

I `git clone https://github.com/plasma-umass/coz.git` and enter the directory. From https://github.com/plasma-umass/coz/issues/186#issuecomment-1003789207 i learn that the ReadMe suggesting

> To build and run one of these benchmarks with coz, just browse to benchmarks/{bench name} and type make bench (or make test for a smaller input size).

is false, at-least for coz/benchmarks. It fails with ` -lcoz-instrumentation`. So i run `cmake -B build -DBUILD_BENCHMARKS=on -DCMAKE_BUILD_TYPE=Debug .` in the top-level repo folder. The dependencies i install one by one. The README says:
> Use latest from https://github.com/aclements/libelfin Version 0.3 does not work

Ubuntu shipped me a `libelfin-dev/jammy,jammy,now 0.3-2.1` will that work? I can't tell. I am annoyed that seemingly *have to* add heavy programs like mysql and other libraries for the examples as there seems to be no way to just build a selection of examples. All the examples build. Good. Now i have a benchmarks/ folder and a build/benchmarks/ folder. The latter contains newly built binaries so let's try that one.

`make bench` in builds/benchmarks does nothing. `make test` fails:
```
Running tests...
Test project /home/joto/coz/build/benchmarks
    Start 1: test_run_kmeans
1/1 Test #1: test_run_kmeans ..................***Failed    0.04 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.04 sec

The following tests FAILED:
          1 - test_run_kmeans (Failed)
Errors while running CTest
Output from these tests are in: /home/joto/coz/build/benchmarks/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
make: *** [Makefile:71: test] Error 8
```
`make test --rerun-failed --output-on-failure` yells at me for using `make` wrong. Guess i have a look at the log then.
```
1/1 Testing: test_run_kmeans
1/1 Test: test_run_kmeans
Command: "/home/joto/coz/benchmarks/check-output.sh" "/home/joto/coz/coz" "run" "---" "/home/joto/coz/build/benchmarks/kmeans/kmeans"
Directory: /home/joto/coz/build
"test_run_kmeans" start time: Aug 11 01:16 CEST
Output:
----------------------------------------------------------
[/home/joto/coz/libcoz/libcoz.cpp:100] bootstrapping coz
[/home/joto/coz/libcoz/libcoz.cpp:128] Including MAIN, which is /home/joto/coz/build/benchmarks/kmeans/kmeans
[/home/joto/coz/libcoz/inspect.cpp:509] Included source file /home/joto/coz/benchmarks/kmeans/kmeans-pthread.c
[/home/joto/coz/libcoz/inspect.cpp:509] Included source file /home/joto/coz/include/coz.h
[/home/joto/coz/libcoz/inspect.cpp:316] Including lines from executable /home/joto/coz/build/benchmarks/kmeans/kmeans
[/home/joto/coz/libcoz/profiler.cpp:75] Starting profiler thread
[/home/joto/coz/libcoz/libcoz.cpp:96] init_coz in progress, do not recurse
[/home/joto/coz/libcoz/profiler.h:123] Thread state not found
Aborted!
  0: libcoz/libcoz.so(_ZN8profiler8on_errorEiP9siginfo_tPv+0xdd) [0x7f24f0786bfd]
  1: /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f24f050e520]
  2: /lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x12c) [0x7f24f0562a7c]
  3: /lib/x86_64-linux-gnu/libc.so.6(raise+0x16) [0x7f24f050e476]
  4: /lib/x86_64-linux-gnu/libc.so.6(abort+0xd3) [0x7f24f04f47f3]
  5: libcoz/libcoz.so(_ZN6ccutil6loggerD1Ev+0x5a) [0x7f24f076e326]
  6: libcoz/libcoz.so(_ZN8profiler21handle_pthread_createEPmPK14pthread_attr_tPFPvS4_ES4_+0x13a) [0x7f24f077c368]
  7: libcoz/libcoz.so(pthread_create+0x3f) [0x7f24f077a676]
  8: libcoz/libcoz.so(+0x84612) [0x7f24f078c612]
  9: libcoz/libcoz.so(_ZN8profiler7startupERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEP4lineib+0x295) [0x7f24f0784d7f]
  10: libcoz/libcoz.so(_Z8init_cozv+0xb7b) [0x7f24f0779ff2]
  11: libcoz/libcoz.so(+0x723fa) [0x7f24f077a3fa]
  12: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f24f04f5d90]
  13: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f24f04f5e40]
  14: libcoz/libcoz.so(coz_libc_start_main+0x7e) [0x7f24f077a56c]
  15: /home/joto/coz/build/benchmarks/kmeans/kmeans(+0x12f5) [0x55ba4331f2f5]
<end of output>
Test time =   0.04 sec
```
I run `build/benchmarks/kmeans/kmeans`. It works. I run `coz run --- build/benchmarks/kmeans/kmeans`. It doesn't the same error repeats over multiple different benchmarks not a single one is working. I find [this thread](https://github.com/plasma-umass/coz/issues/180#issuecomment-1073317104) discussing the same error for Rust. Someone suggests that some dynamic library loading flag is false. *takes note*.

I back out to the root dir and see an coz executeable. I run `./coz run --- build/benchmarks/kmeans/kmeans`. Same error.

### Purge Ubuntu's package
Maybe the ubuntu side install causes problems? I `apt get purge` it away. `find / -name "*coz*" 2> /dev/null` shows all occourances come from the cloned repo. Cleanup by apt was successful. `./coz run --- build/benchmarks/kmeans/kmeans` still fais in the same manner.

### RTLD_NOLOAD
I can't get the preload trick discussed in #180 to work. I remove ` | RTLD_NOLOAD` from
```
static void* get_pthread_handle() {
  if(pthread_handle == NULL && !__atomic_exchange_n(&in_dlopen, true, __ATOMIC_ACQ_REL)) {
    pthread_handle = dlopen("libpthread.so.0", RTLD_NOW | RTLD_GLOBAL | RTLD_NOLOAD);
    __atomic_store_n(&in_dlopen, false, __ATOMIC_RELEASE);
  }

  return pthread_handle;
}
```
and rebuild using `cmake -B build -DBUILD_BENCHMARKS=on -DCMAKE_BUILD_TYPE=Debug` and `make`. Run `./coz run --- build/benchmarks/kmeans/kmeans`. Same error.

I try `sudo make install`. I shudder. No i am not gonna install NodeJS stuff. I do not care about building the viewer. The library, the examples should be seperate repos maybe? I don't know. I feel stuck.

I `ack "RTLD_NOLOAD"` no other concurrences of this string. `cat /etc/lsb-release` says
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.1 LTS"
```
Find no option to not build the viewer. I am stuck and see no path to proceed.

**Comment by emeryberger** [comment]
Thanks for raising this issue. I had not tried to rebuild from source for some time, and clearly things have drifted. My apologies.

I spent yesterday working on this, made some modifications and was able to build everything from the revised instructions now in `README.md` (you should be able to safely leave off nodejs and npm).

You can now cd into `benchmarks`, run `cmake . && make` and it should build all benchmarks successfully.

**Comment by emeryberger** [comment, marked as answer]
This all worked for me on Ubuntu 20, but it's not working on Ubuntu 22 for some reason. I hope to track the issue down soon.

**Reply by emeryberger** [reply]
Works on Ubuntu 22 now.
